framework/src/fwk_status.c

```diff
@@ -11,55 +11,42 @@
 #include <fwk_status.h>
 #include <fwk_macros.h>
 
+static const char * const error_string[] = {
+    [FWK_SUCCESS]        = "SUCCESS",
+    [-FWK_E_PARAM]       = "E_PARAM",
+    [-FWK_E_ALIGN]       = "E_ALIGN",
+    [-FWK_E_SIZE]        = "E_SIZE",
+    [-FWK_E_HANDLER]     = "E_HANDLER",
+    [-FWK_E_ACCESS]      = "E_ACCESS",
+    [-FWK_E_RANGE]       = "E_RANGE",
+    [-FWK_E_TIMEOUT]     = "E_TIMEOUT",
+    [-FWK_E_NOMEM]       = "E_NOMEM",
+    [-FWK_E_PWRSTATE]    = "E_PWRSTATE",
+    [-FWK_E_SUPPORT]     = "E_SUPPORT",
+    [-FWK_E_DEVICE]      = "E_DEVICE",
+    [-FWK_E_BUSY]        = "E_BUSY",
+    [-FWK_E_OS]          = "E_OS",
+    [-FWK_E_DATA]        = "E_DATA",
+    [-FWK_E_STATE]       = "E_STATE",
+    [-FWK_E_INIT]        = "E_INIT",
+    [-FWK_E_OVERWRITTEN] = "E_OVERWRITTEN",
+    [-FWK_E_PANIC]       = "E_PANIC",
+};
+
+static const char * const status_string[] = {
+    [FWK_SUCCESS]        = "SUCCESS",
+    [FWK_PENDING]        = "PENDING",
+};
+
 const char *fwk_status_str(int status)
 {
-    switch (status) {
-    case FWK_PENDING:
-        return "PENDING";
-    case FWK_SUCCESS:
-        return "SUCCESS";
-    case FWK_E_PARAM:
-        return "E_PARAM";
-    case FWK_E_ALIGN:
-        return "E_ALIGN";
-    case FWK_E_SIZE:
-        return "E_SIZE";
-    case FWK_E_HANDLER:
-        return "E_HANDLER";
-    case FWK_E_ACCESS:
-        return "E_ACCESS";
-    case FWK_E_RANGE:
-        return "E_RANGE";
-    case FWK_E_TIMEOUT:
-        return "E_TIMEOUT";
-    case FWK_E_NOMEM:
-        return "E_NOMEM";
-    case FWK_E_PWRSTATE:
-        return "E_PWRSTATE";
-    case FWK_E_SUPPORT:
-        return "E_SUPPORT";
-    case FWK_E_DEVICE:
-        return "E_DEVICE";
-    case FWK_E_BUSY:
-        return "E_BUSY";
-    case FWK_E_OS:
-        return "E_OS";
-    case FWK_E_DATA:
-        return "E_DATA";
-    case FWK_E_STATE:
-        return "E_STATE";
-    case FWK_E_INIT:
-        return "E_INIT";
-    case FWK_E_OVERWRITTEN:
-        return "E_OVERWRITTEN";
-    case FWK_E_PANIC:
-        return "E_PANIC";
-    default:
-        break;
-    }
+    unsigned int error_idx = (unsigned int)(-status);
+
+    if (status >= 0 && status < FWK_ARRAY_SIZE(status_string))
+        return status_string[status];
 
-    if (status < 0)
-        return "unknown-negative";
+    if (status < 0 && error_idx < FWK_ARRAY_SIZE(error_string))
```

Comment:
`else if (error_idx < FWK_ARRAY_SIZE(error_string))`

Reply:
ok

Reply:
actually no:
 if (status positive and index in range) then (return positive status ref.)
 if (status negative and index in range) then (return negative status ref.)
 return "unknwn" ref.

Reply:
Apologies, you're right; that's what I get for doing late Friday reviews!